util/oci/client.go

```diff
@@ -0,0 +1,302 @@
+	argoio "github.com/argoproj/argo-cd/v2/util/io"
+	"github.com/argoproj/argo-cd/v2/util/proxy"
+	"github.com/argoproj/pkg/sync"
+	log "github.com/sirupsen/logrus"
+	"net/http"
+	"net/url"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/file"
+	"oras.land/oras-go/v2/registry/remote"
+	"oras.land/oras-go/v2/registry/remote/auth"
+	"os"
+	"strings"
+	"time"
+)
+
+var (
+	globalLock = sync.NewKeyLock()
+	indexLock  = sync.NewKeyLock()
+)
+
+var _ Client = &nativeOciClient{}
+
+type indexCache interface {
+	SetHelmIndex(repo string, indexData []byte) error
+	GetHelmIndex(repo string, indexData *[]byte) error
+}
+
+// Client is a generic oci client interface
+type Client interface {
+	GetTags(ctx context.Context, noCache bool) (*TagsList, error)
+	ResolveDigest(ctx context.Context, revision string) (string, error)
+	ResolveRevision(ctx context.Context, revision string, noCache bool) (string, error)
+	CleanCache(revision string, project string) error
+	Extract(ctx context.Context, revision string, project string, manifestMaxExtractedSize int64, disableManifestMaxExtractedSize bool) (string, argoio.Closer, error)
+	TestRepo(ctx context.Context) (bool, error)
+}
+
+type Creds struct {
+	Username           string
+	Password           string
+	CAPath             string
+	CertData           []byte
+	KeyData            []byte
+	InsecureSkipVerify bool
+	InsecureHttpOnly   bool
+}
+
+type ClientOpts func(c *nativeOciClient)
+
+func WithIndexCache(indexCache indexCache) ClientOpts {
+	return func(c *nativeOciClient) {
+		c.indexCache = indexCache
+	}
+}
+
+func WithChartPaths(repoCachePaths argoio.TempPaths) ClientOpts {
+	return func(c *nativeOciClient) {
+		c.repoCachePaths = repoCachePaths
+	}
+}
+func NewClient(repoURL string, creds Creds, proxy string, opts ...ClientOpts) (Client, error) {
+	return NewClientWithLock(repoURL, creds, globalLock, proxy, opts...)
+}
+func NewClientWithLock(repoURL string, creds Creds, repoLock sync.KeyLock, proxyUrl string, opts ...ClientOpts) (Client, error) {
+	ociRepo := strings.TrimPrefix(repoURL, "oci://")
+	repo, err := remote.NewRepository(ociRepo)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize repository: %v", err)
+	}
+
+	repo.PlainHTTP = creds.InsecureHttpOnly
+
+	ociUri, err := url.Parse(repoURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse registry host: %v", err)
+	}
+
+	var tlsConf *tls.Config
+	if !repo.PlainHTTP {
+		tlsConf, err = newTLSConfig(creds)
+		if err != nil {
+			return nil, fmt.Errorf("failed setup tlsConfig: %v", err)
+		}
+	}
+
+	var ociRegistry string
+	if ociUri.Port() != "" {
+		ociRegistry = fmt.Sprintf("%s:%s", ociUri.Hostname(), ociUri.Port())
+	} else {
+		ociRegistry = ociUri.Hostname()
+	}
+
+	client := &http.Client{Transport: &http.Transport{
+		Proxy:             proxy.GetCallback(proxyUrl),
+		TLSClientConfig:   tlsConf,
+		DisableKeepAlives: true,
+	}}
+	repo.Client = &auth.Client{
+		Client: client,
+		Cache:  nil,
```

Comment:
Using [`auth.NewCache()`](https://pkg.go.dev/oras.land/oras-go/v2@v2.5.0/registry/remote/auth#NewCache) here can help to reduce the number of authentication requests.

Reply:
I want to avoid the [situation](https://github.com/argoproj/argo-cd/pull/12554#issuecomment-1517462176) which happened in #12554, where using the default cache led to ECR repos failing after 12h. 

Ideally, this should only connect to a remote repo only 12h or so anyway, due to how the repo-server caches artifacts (we will need to ensure that is actually the case in this PR).

Reply:
> where using the default cache led to ECR repos failing after 12h.

Oh from [this comment](https://github.com/argoproj/argo-cd/pull/12554#issuecomment-1473654165) is looks like the error was a `403`, interesting. If the returned error is a `401`, `oras-go` should be able to refresh the token using the provided credentials.

Another note: `auth.DefaultCache` is shared across the process, while `auth.NewCache()` creates a new cache instance.

Anyway, we can leave it as `nil` here if we have concerns. We can come back to improve it if there is performance issue in the future.